[Docs] [Qt] RPC-Console nested commands documentation

lmlsna opened the pull request:
This PR would close issue #9195 by adding documentation for the debug console features (mainly nested commands) which were added in [PR #7783](https://github.com/bitcoin/bitcoin/pull/7783).

The following changes were made to QT debug console code:
- Added a line to the initial message text at the top of the debug console:  

> For more information on using this console type **help-console**.

- Added a pseudo-command `help-console` which is hooked after parsing the request, but before actually executing the RPC thread. It prints the following text to the console as if it were a valid RPC response.

> This console accepts RPC commands using the standard syntax.
>    example:    getblockhash 8
> This console can also accept RPC commands using bracketed syntax.
>    example:    getblockhash(8)
> A space or a comma can be used to separate arguments for either syntax.
>    example:    sendtoaddress \<address\> \<amount\>
>                    sendtoaddress,\<address\>,\<amount\>
> Commands may be nested when specified with the bracketed syntax.
>    example:    getblockinfo(getblockhash(0),true).
> Result values can be queried with a non-quoted string in brackets.
>    example:    getblock(getblockhash(0) true)[height]

This seemed like a reasonably sane way to introduce a fake RPC help command, but

jonasschnelli commented on src/qt/rpcconsole.cpp:
```diff
@@ -392,6 +392,28 @@ void RPCExecutor::request(const QString &command)
+                "   example:    getblockinfo(getblockhash(0),true).\n\n"
+
+                "Result values can be queried with a non-quoted string in brackets.\n"
+                "   example:    getblock(getblockhash(0) true)[height]\n\n" )));
```
maybe add an example that uses access-by-index like `getblock(getblockhash(0) true)[tx][0]`?

lmlsna replied:
Agreed, should have included that.

laanwj replied:
Here the use of "bracket" is correct, referring to the `[]`.

jonasschnelli commented:
Works as intended.
I think it's not prominent enough,... maybe an icon (unsure) or a clickable text (link) in the console text would make more sense.
At least it should auto-complete `help-console` when typing in the prompt line.

lmlsna commented:
I figured that it made sense (at least thematically 😅) for console help to be available from the inside the console, but that's basically just my personal preference. I'm guessing the subset of users interested in nested rpc commands are probably more command-line-centric that then the average user. Though a clickable link in the first bit of text might well be a better/cleaner solution. 

If kept as a command, I for sure think it should be added to the autocomplete list. I can add that in as well, I just didn't want to go overboard before getting some feedback.

If implemented as a button/link, what should it reference/do? Pop a new window with text (like help > commands) in the file menu?

laanwj commented:
Thanks for doing this! will test.

laanwj commented:
> I figured that it made sense (at least thematically 😅) for console help to be available from the inside the console,

Agree, I kind of like it as a command. An icon could be added later, but this is what I had in mind. Only have a few comments on the text.

laanwj commented on src/qt/rpcconsole.cpp:
```diff
@@ -392,6 +392,28 @@ void RPCExecutor::request(const QString &command)
+                "This console accepts RPC commands using the standard syntax.\n"
+                "   example:    getblockhash 8\n\n"
+
+                "This console can also accept RPC commands using bracketed syntax.\n"
```
This is `parenthesized` not `bracketed` syntax, I'd say?

laanwj commented on src/qt/rpcconsole.cpp:
```diff
@@ -392,6 +392,28 @@ void RPCExecutor::request(const QString &command)
+                "This console can also accept RPC commands using bracketed syntax.\n"
+                "   example:    getblockhash(8)\n\n"
+
+                "A space or a comma can be used to separate arguments for either syntax.\n"
```
Maybe "delimit" instead of "separate"?

laanwj commented on src/qt/rpcconsole.cpp:
```diff
@@ -747,10 +769,12 @@ void RPCConsole::clear(bool clearHistory)
     message(CMD_REPLY, (tr("Welcome to the %1 RPC console.").arg(tr(PACKAGE_NAME)) + "<br>" +
                         tr("Use up and down arrows to navigate history, and %1 to clear screen.").arg("<b>"+clsKey+"</b>") + "<br>" +
-                        tr("Type <b>help</b> for an overview of available commands.")) +
+                        tr("Type <b>help</b> for an overview of available commands.<br>") +
```
Please keep the `<br>` out of the translation message, as it frequently gets lost by translators.

lmlsna replied:
I totally agree with all of this.

lmlsna commented:
I have the autocomplete working and will make the suggested changes to the help text. I'll push a revision today or tomorrow. Thanks for the input guys!

jonasschnelli commented:
Thanks! Looks good!
utACK after [squash](https://stackoverflow.com/questions/5189560/squash-my-last-x-commits-together-using-git).

lmlsna commented:
squished

promag commented:
utACK 5305dda.

jonasschnelli commented:
utACK 5305dda4c3f2d0b9d6a1bf9d5c4cc1c6215ed475

MeshCollider commented on src/qt/rpcconsole.cpp:
```diff
@@ -392,11 +392,37 @@ void RPCExecutor::request(const QString &command)
+                "Named result objects can be queried with a non-quoted key string.\n"
+                "   example:    getblock(getblockhash(0) true)[tx]\n\n"
+
+                "Numeric result objects can be queried with an index integer in brackets.\n"
```
I'm not too sure about calling them "numeric result objects", numeric just sounds like it's a type of number, rather than an unnamed list/array or something

lmlsna replied:
Yeah, good point. You think changing those last 2 lines to just "Results can be queried with..." is better?

lmlsna replied:
I'll change that last bit and resquash with a better commit message.

meshcollider reviewed (commented):
utACK https://github.com/bitcoin/bitcoin/commit/5305dda4c3f2d0b9d6a1bf9d5c4cc1c6215ed475

Commit message needs modification though,  perhaps "Add help-console command to Qt debug console" or something

lmlsna committed:
Add help-console command to Qt debug console

     - Added `help-console` to the list of autocompletion strings
     - Implemented requested changes to help message:
       - Added an example that uses access-by-index `getblock(getblockhash(0) true)[tx][0]`
       - Replace "bracketed syntax" to "parenthesized syntax" where applicable
       - Replace "separate" with "delimit"
       - Removed `<br>` and `<b>help/help-console</b>` from translation strings, since these parts don't change between languages
       - Changed examples to be based off `getblock 0` so they will work even with pruned/no blockchain and `disablewallet` if copied and pasted
       - Clarified syntax for queries of named/unnamed result objects.

MeshCollider commented:
re-utACK https://github.com/bitcoin/bitcoin/pull/11698/commits/c3055bbea1255cd791564b855b6951a9f4c510f9

MarcoFalke commented on src/qt/rpcconsole.cpp:
```diff
@@ -392,11 +392,37 @@ void RPCExecutor::request(const QString &command)
         std::string executableCommand = command.toStdString() + "\n";
+
+        // Catch the console-only-help command before RPC call is executed and reply with help text as-if a RPC reply.
+        if(executableCommand == "help-console\n")
```
nit: Should use `.starts_with("help-console")` to ignore trailing white space. Though, can wait, since that seems to be part of cpp20

maflcko reviewed (commented):
post merge ACK c3055bbea1255cd791564b855b6951a9f4c510f9
